Add initial Usage sketch. (#1)
did-coop/wallet-attached-storage: README.md

See in progress spec: https://did-coop.github.io/wallet-attached-storage-spec/

### App Identity and Key Management

To use Wallet Attached Storage (to provision data spaces, to create collections,
and to read and write resources), the application or service using this client
will need its own cryptographically provable identity (in the form of a
[DID](https://w3c.github.io/did-core/)).

At minimum, this means that the app or service using this client will need to
create and manage its own public-private key pair. The challenges and security
considerations of cryptographic key management are beyond the scope of this
usage document, but roughly:

* For in browser **client-only Single Page Applications** (SPAs), the app's
  identity is ephemeral -- essentially a key pair will be generated for each

>   new user session, and stored in the browser's Local Storage.

Reviewer: ```suggestion
  new user session, and stored in the browser.
```

Mentioning "Local Storage" reads like it is a good idea to use `globalThis.localStorage`, when there are better storage options like non-exportable key in indexedDb. We should either not recommend a particular thing like "Local Storage" or include IndexedDB in the recommendations. This does the former.

Author: Agreed yeah. I just meant the general idea, not specific mechanism.

## Usage

This client assumes:

* a remote Wallet Attached Storage server is available
* the app is able to create and store a public-private key pair, see the
  [App Identity and Key Management](#app-identity-and-key-management) section
  for more details.

Example key pair generation (for an in-browser SPA), for persistence in local
storage.

```ts
import { Ed25519Signer } from '@did.coop/did-key-ed25519'

const appDidSigner = await Ed25519Signer.generate()
```

Example storing and loading:

```ts
// Serialize to JSON for storage
const exportedKeyPair = appDidSigner.toJSON()

> localStorage.setItem('app-DID', JSON.stringify(exportedKeyPair))

Reviewer: Hmm. Ideally these examples would store non exportable secret keys.

Unfortunately, that is only possible for ed25519 keys on safari and firefox atm. chromium 🔜  https://github.com/WICG/webcrypto-secure-curves/issues/20

https://caniuse.com/mdn-api_subtlecrypto_sign_ed25519

I don't mind leaving this for now, but I'm curious what you think of:
* we should support secp256r1 keys/sigs not only ed25519. iiuc these are FIPS approved and supported across browsers via WebCrypto
* then we can maybe have some kind of generic WebCrypto signer that uses the most secure curve supported by the user agent, all non-exportable

Author: Totally agree, re non-exportable keys. Unfortunately we're just supporting ed25519 for the first iteration, so I figured for the first spike, we can do localstorage.

// Load from storage and turn back into a DID Signer
const loadedKeyPair = localStorage.getItem('app-DID')
const appDidSigner = Ed25519Signer.fromJSON(loadedKeyPair)
```

Create a Wallet Attached Storage Client:

```ts

> import { WasClient } from '@did-coop/wallet-attached-storage'

Reviewer: ```suggestion
import { WasClient } from '@did-coop/wallet-attached-storage'
```

Can we export a more explicit name like `WalletAttachedStorageClient` or just `WalletStorageClient`?
IMO it makes it all more intelligble and easy to grok if we avoid 'Was', which I think folks won't immediately recognize the meaning of.

Author: Good point, will change

const config = {
  was: {
    url: 'https://data.pub'
  }
}

const storage = new WasClient({ url: config.was.url, signer: appDidSigner })
```

Provision (create and set up) a Data Space (this performs a [Create Space HTTP
API operation](https://did-coop.github.io/wallet-attached-storage-spec/#create-space-operation)):

```ts
let space
try {

>   space = await storage.space('demo-space')

Reviewer: This is similar to the fetch-client API, but slightly different. e.g. that one has `pub.space()` but returns synchronously

Reviewer: user will have zcap to space, which will contain a space id (urn:uuid), and they can create a space from that like `storage.space(uuid)`

>   space = await storage.space('demo-space')

Reviewer: ```suggestion
  space = await storage.space()
```

I don't think we should take `name` here.

} catch (e) {
  console.error('Error initializing data space:', e)
  throw e
}
```

Now you can read and write resources to and from collections:

```ts
// Create a request to list the contents of the 'credentials' collection
const cursor = space.collection('credentials').list()

// Iterate through the results
for await (const vc of cursor) {
  console.log(vc);
}
// or cast the results to an array

> const allItems = await cursor.toArray()

Reviewer: I don't think we should have this method tbh. These collections can get really big and imho we should only epose streaming/async iterfaces.

if cursor is AsyncIterable, anyone can already do this via const items = await Array.fromAsync(cursor)

Author: Sounds good.

// Write a VC to a collection

> const { url } = await space.collection('credentials').put(myVc)

Reviewer: This is confusingly different from @data.pub/fetch-client, where this would PUT a representation of the collection... not append an item to a collection...
